### PR TITLE
TDL-17514: Implement request timeout

### PR DESCRIPTION
**Pull request by savan-chovatiya**
# Description of change
[TDL-17514](https://jira.talendforge.org/browse/TDL-17514): Implement request timeout
- Added request timeout for API requests with default timeout 300 seconds
**NOTE**: the build is failing due to integration tests failure but that is fixed and all tests are written as part of #5 

# Manual QA steps
 - Set small timeout and verified that sync calls are backing off for timeout error.
- Provided timeout in integer, float, and string format and verified that it's used in the request.

 
# Risks
 - 
 
# Rollback steps
 - revert this branch

### CODE REVIEW

File: tests/unittests/test_request_timeout.py [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+        # Call _make_request method which call Session.request with timeout
+        client._make_request("http://test", "test", "test")
+
+        # Verify session.request is called with expected timeout
+        args, kwargs = mocked_request.call_args
+        self.assertEqual(kwargs.get('timeout'), 300) # Verify timeout argument
+
+    def test_integer_request_timeout_in_config(self, mocked_request):
+        """
+            Verify that if request_timeout is provided in config(integer value) then it should be use
```
Comment (dbshah1212) [inline tests/unittests/test_request_timeout.py:32]:
```suggestion
            Verify that if request_timeout is provided in config(integer value) then it should be used.
```

File: tests/unittests/test_request_timeout.py [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+        # Call _make_request method which call Session.request with timeout
+        client._make_request("http://test", "test", "test")
+
+        # Verify session.request is called with expected timeout
+        args, kwargs = mocked_request.call_args
+        self.assertEqual(kwargs.get('timeout'), 100.0) # Verify timeout argument
+
+    def test_float_request_timeout_in_config(self, mocked_request):
+        """
+            Verify that if request_timeout is provided in config(float value) then it should be use
```
Comment (dbshah1212) [inline tests/unittests/test_request_timeout.py:45]:
```suggestion
            Verify that if request_timeout is provided in config(float value) then it should be used.
```

File: tests/unittests/test_request_timeout.py [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+        # Call _make_request method which call Session.request with timeout
+        client._make_request("http://test", "test", "test")
+
+        # Verify session.request is called with expected timeout
+        args, kwargs = mocked_request.call_args
+        self.assertEqual(kwargs.get('timeout'), 100.5) # Verify timeout argument
+
+    def test_string_request_timeout_in_config(self, mocked_request):
+        """
+            Verify that if request_timeout is provided in config(string value) then it should be use
```
Comment (dbshah1212) [inline tests/unittests/test_request_timeout.py:58]:
```suggestion
            Verify that if request_timeout is provided in config(string value) then it should be used.
```

File: tests/unittests/test_request_timeout.py [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+        # Call _make_request method which call Session.request with timeout
+        client._make_request("http://test", "test", "test")
+
+        # Verify session.request is called with expected timeout
+        args, kwargs = mocked_request.call_args
+        self.assertEqual(kwargs.get('timeout'), 100) # Verify timeout argument
+
+    def test_empty_string_request_timeout_in_config(self, mocked_request):
+        """
+            Verify that if request_timeout is provided in config with empty string then default value is used
```
Comment (dbshah1212) [inline tests/unittests/test_request_timeout.py:71]:
```suggestion
            Verify that if request_timeout is provided in the config with an empty string then the default value got used.
```

File: tests/unittests/test_request_timeout.py [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+        # Call _make_request method which call Session.request with timeout
+        client._make_request("http://test", "test", "test")
+
+        # Verify session.request is called with expected timeout
+        args, kwargs = mocked_request.call_args
+        self.assertEqual(kwargs.get('timeout'), 300) # Verify timeout argument
+
+    def test_zero_request_timeout_in_config(self, mocked_request):
+        """
+            Verify that if request_timeout is provided in config with zero value then default value is used
```
Comment (dbshah1212) [inline tests/unittests/test_request_timeout.py:84]:
```suggestion
            Verify that if request_timeout is provided in the config with zero value then the default value got used.
```

File: tests/unittests/test_request_timeout.py [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+        # Call _make_request method which call Session.request with timeout
+        client._make_request("http://test", "test", "test")
+
+        # Verify session.request is called with expected timeout
+        args, kwargs = mocked_request.call_args
+        self.assertEqual(kwargs.get('timeout'), 300) # Verify timeout argument
+
+    def test_zero_string_request_timeout_in_config(self, mocked_request):
+        """
+            Verify that if request_timeout is provided in config with zero in string format then default value is used
```
Comment (dbshah1212) [inline tests/unittests/test_request_timeout.py:97]:
```suggestion
            Verify that if request_timeout is provided in the config with zero in string format then the default value got used.
```

File: tests/unittests/test_request_timeout.py [hunk 7]
```diff
@@ -0,0 +1,123 @@
...
+        self.assertEqual(kwargs.get('timeout'), 300) # Verify timeout argument
+
+
+@mock.patch("time.sleep")
+class TestRequestTimeoutBackoff(unittest.TestCase):
+
+    @mock.patch('requests.Session.send', side_effect = requests.exceptions.Timeout)
+    def test_request_timeout_backoff(self, mocked_request, mocked_sleep):
+        """
+            Verify request function is backoff for 3 times on Timeout exception
```
Comment (dbshah1212) [inline tests/unittests/test_request_timeout.py:115]:
```suggestion
            Verify request function is backing off 3 times on the Timeout exception.
```

File: tests/unittests/test_client.py [hunk 8]
```diff
@@ -3,7 +3,7 @@
 
 @pytest.fixture
 def client():
-    return SailthruClient('test', 'test', 'test')
+    return SailthruClient('test', 'test', 'test', 300)
```
Comment (dbshah1212) [inline tests/unittests/test_client.py:6]: Will this impact existing test cases?

Comment (savan-chovatiya) [reply tests/unittests/test_client.py:6]: The extra parameter is added in the `SailthruClient` for the request_timeout so needs to pass it in SailthruClient initialization but existing test cases are not for request timeout so other assertions will be passed.

File: tests/unittests/test_request_timeout.py (hunk 1 above)
Comment (savan-chovatiya) [reply tests/unittests/test_request_timeout.py]: Updated comment.

File: tests/unittests/test_request_timeout.py (hunk 2 above)
Comment (savan-chovatiya) [reply tests/unittests/test_request_timeout.py]: Updated comment.

File: tests/unittests/test_request_timeout.py (hunk 4 above)
Comment (savan-chovatiya) [reply tests/unittests/test_request_timeout.py]: Updated comment.

File: tests/unittests/test_request_timeout.py (hunk 5 above)
Comment (savan-chovatiya) [reply tests/unittests/test_request_timeout.py]: Updated comment.

File: tests/unittests/test_request_timeout.py (hunk 6 above)
Comment (savan-chovatiya) [reply tests/unittests/test_request_timeout.py]: Updated comment.

File: tests/unittests/test_request_timeout.py (hunk 7 above)
Comment (savan-chovatiya) [reply tests/unittests/test_request_timeout.py]: Updated comment.

**Review by dbshah1212 — APPROVED** [review]

**Review by karanpanchal-crest — APPROVED** [review]

**Review by KrisPersonal — APPROVED** [review]

**Review by kspeer825 — APPROVED** [review]

**Pull request merged**